x-pack/plugins/features/server/feature_registry.ts

```diff
@@ -169,19 +190,106 @@ export class FeatureRegistry {
+          );
+        }
+
+        if (!isFeatureDeprecated && privilege.replacedBy) {
+          throw new Error(
+            `Feature "${feature.id}" is not deprecated and must not define a "replacedBy" property for privilege "${privilegeId}".`
+          );
+        }
+
+        const replacedByReferences = privilege.replacedBy
+          ? 'default' in privilege.replacedBy
+            ? [...privilege.replacedBy.default, ...privilege.replacedBy.minimal]
+            : privilege.replacedBy
+          : [];
+        for (const featureReference of replacedByReferences) {
+          const referencedFeature = this.kibanaFeatures[featureReference.feature];
+          if (!referencedFeature) {
+            throw new Error(
+              `Cannot replace privilege "${privilegeId}" of deprecated feature "${feature.id}" with privileges of feature "${featureReference.feature}" since such feature is not registered.`
+            );
+          }
+
+          if (referencedFeature.deprecated) {
+            throw new Error(
+              `Cannot replace privilege "${privilegeId}" of deprecated feature "${feature.id}" with privileges of feature "${featureReference.feature}" since the referenced feature is deprecated.`
+            );
+          }
+
+          // Collect all known feature and sub-feature privileges for the referenced feature.
+          const knownPrivileges = new Map(
+            collectPrivileges(referencedFeature).concat(
+              collectSubFeaturesPrivileges(referencedFeature)
+            )
+          );
+
+          for (const privilegeReference of featureReference.privileges) {
+            const referencedPrivilege = knownPrivileges.get(privilegeReference);
+            if (!referencedPrivilege) {
+              throw new Error(
+                `Cannot replace privilege "${privilegeId}" of deprecated feature "${feature.id}" with privilege "${privilegeReference}" of feature "${featureReference.feature}" since such privilege is not registered.`
+              );
+            }
+
+            if (referencedPrivilege.disabled) {
+              throw new Error(
+                `Cannot replace privilege "${privilegeId}" of deprecated feature "${feature.id}" with disabled privilege "${privilegeReference}" of feature "${featureReference.feature}".`
+              );
+            }
+          }
+        }
+      }
     }
+  }
 
-    let features = Object.values(this.kibanaFeatures);
+  public getAllKibanaFeatures({
+    license,
+    ignoreLicense = false,
+    omitDeprecated = false,
+  }: GetKibanaFeaturesParams = {}): KibanaFeature[] {
+    if (!this.locked) {
+      throw new Error('Cannot retrieve Kibana features while registration is still open');
+    }
 
     const performLicenseCheck = license && !ignoreLicense;
+    const features = [];
+    for (const feature of Object.values(this.kibanaFeatures)) {
```

Comment:
__note:__ changed the logic to make sure we loop over all features only once while filtering and mapping.